moshi/src/main/java/com/squareup/moshi/JsonAdapter.java

```diff
@@ -103,6 +103,30 @@ public final String toJson(T value) {
     };
   }
 
+  /**
+   * Returns a JSON adapter equal to this, but that throws a {@link JsonDataException} when
+   * {@linkplain JsonReader#setFailOnUnknown(boolean) unknown values} are encountered. This
+   * constraint applies to both the top-level message handled by this type adapter as well as to
+   * nested messages.
+   */
+  public final JsonAdapter<T> failOnUnknown() {
+    final JsonAdapter<T> delegate = this;
```

Comment:
Instead of failing, you should consider using a callback for the programmer to handle it.